gst/nnstreamer/elements/gsttensor_trainer.c

```diff
@@ -0,0 +1,1182 @@
+    const GValue * value);
+static void gst_tensor_trainer_set_prop_input_dimension (GstTensorTrainer *
+    trainer, const GValue * value);
+static void gst_tensor_trainer_set_prop_input_type (GstTensorTrainer * trainer,
+    const GValue * value);
+static void gst_tensor_trainer_find_framework (GstTensorTrainer * trainer,
+    const char *name);
+static void gst_tensor_trainer_create_framework (GstTensorTrainer * trainer);
+static gsize gst_tensor_trainer_get_tensor_size (GstTensorTrainer * trainer,
+    guint index, gboolean is_input);
+static void gst_tensor_trainer_create_model (GstTensorTrainer * trainer);
+static void gst_tensor_trainer_train_model (GstTensorTrainer * trainer);
+static void gst_tensor_trainer_output_dimension (GstTensorTrainer * trainer);
+static void gst_tensor_trainer_output_type (GstTensorTrainer * trainer);
+
+/**
+ * @brief initialize the tensor_trainer's class
+ */
+static void
+gst_tensor_trainer_class_init (GstTensorTrainerClass * klass)
+{
+  GObjectClass *gobject_class;
+  GstElementClass *gstelement_class;
+
+  GST_DEBUG_CATEGORY_INIT (GST_CAT_DEFAULT, "tensor_trainer", 0,
+      "Tensor trainer to train neural network model");
+
+  gobject_class = G_OBJECT_CLASS (klass);
+  gstelement_class = GST_ELEMENT_CLASS (klass);
+
+  gobject_class->set_property =
+      GST_DEBUG_FUNCPTR (gst_tensor_trainer_set_property);
+  gobject_class->get_property =
+      GST_DEBUG_FUNCPTR (gst_tensor_trainer_get_property);
+  gobject_class->finalize = GST_DEBUG_FUNCPTR (gst_tensor_trainer_finalize);
+
+  /* Called when the element's state changes */
+  gstelement_class->change_state =
+      GST_DEBUG_FUNCPTR (gst_tensor_trainer_change_state);
+
+  /* Install properties for tensor_trainer */
+  g_object_class_install_property (gobject_class, PROP_FRAMEWORK,
+      g_param_spec_string ("framework", "Framework", "Neural network framework",
+          DEFAULT_PROP_FRAMEWORK,
+          G_PARAM_READWRITE | GST_PARAM_MUTABLE_READY |
+          G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_MODEL_CONFIG,
+      g_param_spec_string ("model-config", "Model configuration file path",
+          "Configuration file path for creating model",
+          DEFAULT_STR_PROP_VALUE,
+          G_PARAM_READWRITE | GST_PARAM_MUTABLE_READY |
+          G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_MODEL_SAVE_PATH,
+      g_param_spec_string ("model-save-path", "Model save path",
+          "File path to save the model", DEFAULT_STR_PROP_VALUE,
+          G_PARAM_READWRITE | GST_PARAM_MUTABLE_READY |
+          G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_INPUT,
+      g_param_spec_string ("input", "Input dimension",
+          "Input tensors dimension from inner array, up to 4 dimensions ?",
+          DEFAULT_STR_PROP_VALUE,
+          G_PARAM_READWRITE | GST_PARAM_MUTABLE_READY |
+          G_PARAM_STATIC_STRINGS));
```

Comment:
input dimensions are optional, right?

Reply:
Thank you.
If we receive it as an optional, we need to pass the value to the sub plugin after parsing the caps in tensor_trainer.
I can change it anytime.